댓글 기능 구현

Nitoriubar opened the pull request:
## 📝 세부사항
- CRUD 구현
- 중복되는 코드 리팩토링 진행함
- `GetReplyResponse`에 하위의 값들을 담아 보내는 경우, 값이 존재하지 않으면 null, 존재하면 객체를 보내야하는 점에 있어 삼항연산자로 구현함 
  - `parentReplyId`
  - `subReplyUserName`
  - `replyCnt`

## 📗 참고사항
- [Get방식에서는 RequestBody를 사용할 수 없음](https://lazyfriend.hashnode.dev/get-request-body)

## 📎 관련 이슈
- Close #33

Nitoriubar committed:
feat: 댓글 기능 뼈대 구현

- ReplyController 생성
- Reply 수정
- ReplyService 생성
- CreateReplyRequest 생성 후 작성완료
- ReplyRepository 생성
- ReplyResponse 생성
- ReplyService에 댓글 생성 구현 중

Nitoriubar committed:
feat: 댓글 조회 기능 구현

Nitoriubar committed:
feat: 댓글 수정 및 삭제 기능 구현

- 중복되는 코드가 있어 Refactoring 진행함

Nitoriubar committed:
style: 통일한 메서드 이름을 갖추기 위해 수정함

- delete -> deleteReply

sangrimlee commented on src/main/java/com/triprint/backend/domain/reply/controller/ReplyController.java:
```diff
@@ -0,0 +1,70 @@
+	@PostMapping()
+	@PreAuthorize("hasRole('ROLE_USER')")
+	public ResponseEntity<ReplyResponse> createReply(
+		@Valid @RequestBody ReplyRequest replyRequest,
```
`Trip`에서 사용하였던 것 처럼 Prefix를 붙여서 명시적으로 만들어주면 더 좋을 것 같아요!

Nitoriubar replied:
넵! 수정하겠습니다 :)

sangrimlee commented on src/main/java/com/triprint/backend/domain/reply/dto/ReplyRequest.java:
```diff
@@ -0,0 +1,19 @@
+public class ReplyRequest {
+
+	@Nullable
+	private Long subReplyUserId;
+
+	private String contents;
+
+	private Long postId;
+
+	@Nullable
+	private Long parentReplyId;
```
지금은 생성할 때랑 수정할 때 모두 `ReplyRequest`가 사용되고 있는데, 댓글 수정에 `postId`, `parentReplyId`, `subReplyUserId`를 모두 수정할 수 있도록 할 것인지 고민해봐야 할 것 같아요 😃

sangrimlee commented on src/main/java/com/triprint/backend/domain/reply/service/ReplyService.java:
```diff
@@ -0,0 +1,119 @@
+			throw new ResourceNotFoundException("해당하는 게시물이 존재하지 않습니다.");
+		});
+
+		if (replyRequest.getParentReplyId() != null) {
+			parentReply = replyRepository.findById(replyRequest.getParentReplyId()).orElseThrow(() -> {
+				throw new ResourceNotFoundException("해당하는 댓글이 존재하지 않습니다.");
+			});
+		}
+		if (replyRequest.getSubReplyUserId() != null) {
+			subReplyUser = userRepository.findById(replyRequest.getSubReplyUserId()).orElseThrow(() -> {
+				throw new ResourceNotFoundException("일치하는 User가 존재하지 않습니다.");
+			});
+		}
```
이렇게 처리하면 `parentReplyId`랑 `getSubReplyUserId` 둘 중 하나만 `null`이 되는 경우가 있을 수 있는데, 이러한 경우에는 정상적이지 않은 요청이라는 생각이 드는데, 이에 대한 Handling도 필요해보여요!

sangrimlee commented on src/main/java/com/triprint/backend/domain/reply/service/ReplyService.java:
```diff
@@ -0,0 +1,119 @@
+	private ReplyDto isExist(ReplyRequest replyRequest) {
+		User subReplyUser = null;
+		Reply parentReply = null;
+		Post post = postRepository.findById(replyRequest.getPostId()).orElseThrow(() -> {
+			throw new ResourceNotFoundException("해당하는 게시물이 존재하지 않습니다.");
+		});
```
사실 이러한 로직이 자주 발생하고 있는데, 매번 에러 메시지를 명시하는 것보다, 미리 에러 메시지를 명시해두고 가져다 쓰거나, 에러코드를 활용해야 할 것 같아요.

그리고, 이러한 로직을 서비스 단으로 분리해 보는 것도 괜찮을 것 같기도 해요. 이건 Spring Boot에서 어떤 식으로 해결하는지 찾아보면 좋을 것 같습니다.

sangrimlee reviewed (commented):
고생하셨습니다! 이번엔 리뷰가 몇 개 있는데, 고민해 볼 만한 것들을 끄적여봤습니다. 같이 고민해 봐요😄

Nitoriubar commented:
> 고생하셨습니다! 이번엔 리뷰가 몇 개 있는데, 고민해 볼 만한 것들을 끄적여봤습니다. 같이 고민해 봐요😄

너무 좋아요! 항상 고맙습니다! :)

Nitoriubar committed:
style: 매서드에 따른 Request변수 명칭 변경함

Nitoriubar committed:
fix: 피드백 반영

- 1,2: Prefix -> DTO를 각각  나누어 명시적으로 변경함
- 3: 분기처리하여 변경함
- 4: ErrorMessage클래스를 만들어 변경함(전반적인 Error를 핸들링하는
  클래스에 매개변스로  String을 받게 되어 있어 변경할 경우, 연관된 모든
  클래스를 수정해야함으로 이와 변경함)

Nitoriubar committed:
fix : 사용하지 않는 Enum 삭제 및 서비스 변경사항에 맞게 getReplyResponseDto수정
